File: problem_1/solution.py
#!/usr/bin/env python2.7

import unittest

MULTIPLES = [ 3, 5, 6, 9, 10, 12, 15, 18, 20 ]

class SolutionTest(unittest.TestCase):
    def test_next_multiple ( self ):
        solution = Solution()

        for n in MULTIPLES: self.assertEqual(n, solution.next())


    def test_solve_for ( self ):
        self.assertEquals(3, Solution.solve_for(5))

        self.assertEquals(23, Solution.solve_for(10))

        self.assertEquals(78, Solution.solve_for(20))


class Solution(object):
    def __init__ ( self ):
        self.a = 0; self.b = 0


    def next ( self ):
        a, b = self.a, self.b

        self.a += a <= b and 3

        self.b += b <= a and 5

        return self.a < self.b and self.a or self.b


    @classmethod
    def solve_for ( cls, limit ):
        i = cls(); n = i.next(); sum = 0

        while ( n < limit ):
            sum += n; n = i.next()

        return sum


if __name__ == '__main__':
    limit = 1000

    print 'SOLUTION -- Find the sum of all the multiples of 3 or 5 below %d: %d\n\n' % (
        limit, Solution.solve_for(limit)
    )

    unittest.main()

File: problem_1/solution.rb
#!/usr/bin/env ruby

require 'test/unit'

$MULTIPLES = [ 3, 5, 6, 9, 10, 12, 15, 18, 20 ]

class SolutionTest < Test::Unit::TestCase
    def test_next_multiple
        solution = Solution.new

        $MULTIPLES.each do |n|
            assert_equal n, solution.next
        end
    end


    def test_solve_for
        assert_equal 3, Solution.solve_for(5)

        assert_equal 23, Solution.solve_for(10)

        assert_equal 78, Solution.solve_for(20)
    end
end


class Solution
    def initialize
        @a = 0; @b = 0
    end


    def next
        a, b = @a, @b

        @a += 3 if a <= b

        @b += 5 if b <= a

        @a < @b and @a or @b
    end


    def self.solve_for limit
        i = self.new; n = i.next; sum = 0

        while n < limit do
            sum += n; n = i.next
        end

        return sum
    end
end #Solution

limit = 1000

print "SOLUTION -- Find the sum of all the multiples of 3 or 5 below %d: %d\n\n" % [
    limit, Solution.solve_for(limit)
]

